File: app/src/main/kotlin/io/htql/Htql.kt
package io.htql

import io.htql.parser.HtqlLexer
import io.htql.parser.HtqlParser
import io.htql.visitor.HtqlRuntimeVisitor
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.io.File
import kotlin.system.exitProcess


val urlPattern = Regex("(http|https)://[a-zA-Z0-9_.-]+(:[0-9]+)?(/[a-zA-Z0-9_.-/?&%#=]*)?")

fun main() {
    println("welcome to HTQL - HYPER TEXT QUERY LANGUAGE")

    println("Please enter your HTQL query:")

    val expression = readlnOrNull() ?: exitProcess(1)

    evaluateDocument(expression) { document ->
        val lexer = HtqlLexer(CharStreams.fromString(expression))
        val tokens = CommonTokenStream(lexer)
        val parser = HtqlParser(tokens)
        val tree = parser.query()


        val visitor = HtqlRuntimeVisitor(document = document)
        val nodes = visitor.visit(tree)

        println(nodes)
    }
}

fun evaluateDocument(expression: String, block: (document: Document) -> Unit) {
    val urlMatch = urlPattern.find(expression)
    if (urlMatch != null) {
        val url = urlMatch.value
        block(Jsoup.connect(url).get())
    } else {
        block(Jsoup.parse(File("app/src/main/resources/test.html")))
    }
}

File: golang/htql.go
package main

import (
	"flag"
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"github.com/antlr4-go/antlr/v4"
	htql "htql/htql_parser_go"
	parser2 "htql/parser"
	"log"
	"net/http"
	"os"
	"regexp"
)

var urlPattern = regexp.MustCompile(`(http|https)://[a-zA-Z0-9_.-]+(:[0-9]+)?(/[a-zA-Z0-9_.-/?&%#=]*)?`)

func main() {
	expression := flag.String("query", "", "HTQL query")

	flag.Parse()

	evaluateDocument(*expression, func(document *goquery.Document) {

		input := antlr.NewInputStream(*expression)
		lexer := htql.NewHtqlLexer(input)
		tokens := antlr.NewCommonTokenStream(lexer, antlr.TokenDefaultChannel)
		parser := htql.NewHtqlParser(tokens)

		tree := parser.Query()

		visitor := parser2.NewHtqlRuntimeVisitor(document)
		nodes := visitor.Visit(tree)

		fmt.Println(nodes)
	})
}

func evaluateDocument(expression string, block func(document *goquery.Document)) {
	if urlMatch := urlPattern.FindString(expression); urlMatch != "" {

		res, err := http.Get(urlMatch)
		if err != nil {
			log.Fatal(err)
		}

		defer res.Body.Close()
		if res.StatusCode != 200 {
			log.Fatalf("status code error: %d %s", res.StatusCode, res.Status)
		}

		doc, err := goquery.NewDocumentFromReader(res.Body)
		if err != nil {
			log.Fatal(err)
		}

		block(doc)
	} else {
		file, err := os.Open("./test.html")
		if err != nil {
			log.Fatalf("Failed to open file: %v", err)
		}
		defer file.Close()

		doc, err := goquery.NewDocumentFromReader(file)
		if err != nil {
			log.Fatalf("Failed to parse HTML file: %v", err)
		}
		block(doc)
	}
}
